### PR TITLE
Feature/odl resources management

### PR DESCRIPTION
**Pull request by adrian-rosello**
This pull request introduces implementation of the IRootResourceProvider capability for the ODL controller, but also fixes some bugs discovered when testing the implementation with a real ODL instance.

The bugs solved in this pull requests are related to the ClientProvider
* **APIClientProviderAdapter** did not get the credentials form the RootResourceDescriptor.
* The **InternalCXFClientProvider** did not set username/passwords credentials properly. They were set in the bean after the client was created, so it had no effect on the client. We could not move this code before the client creation, so the client was once again created to override the configuration with the **HttpBasicAuthentication** information.

Additionally, this pull request implements the **IRootResourcProvider** capability for the ODL controller. Once this capability is activated, it uses the ODL client to get all its instantiated devices, and create (for each of them) an **IRootResource** instance representing those devices. These new resources are managed by this capability.

Pull request fixes following task: http://jira.i2cat.net/browse/OPENNAAS-1579

### CODE REVIEW

File: extensions/odl/capabilities/src/main/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProvider.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+		for (NodeProperties nodeProperties : nodes.getNodeProperties()) {
+			// get information from the node
+			String nodeId = nodeProperties.getNode().getNodeID();
+			Type nodeType = parseNodeType(nodeProperties.getNode().getNodeType());
+
+			// initialize new RootResource representing this node.
+			Collection<Endpoint> odlResourceEndpoints = new ArrayList<Endpoint>(resource.getDescriptor().getEndpoints());
+			IRootResource odlResource = new RootResource(RootResourceDescriptor.create(new Specification(nodeType), odlResourceEndpoints));
+			rmListener.resourceAdded(odlResource, this, IRootResourceProvider.class);
```
Comment (isartcanyameres) [inline extensions/odl/capabilities/src/main/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProvider.java]: I'm not sure whether resources should be added to IRootResourceProvider or to IRootResourceAdministration instead.
Are you sure resources of other types are added to IRootResourceProvider?
(Just a question :))

Comment (adrian-rosello) [inline extensions/odl/capabilities/src/main/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProvider.java]: It depends. "Normal" networks contain an IRootResourceAdministration implementation managing this capabilities.
But in special cases (NITOS, ODL) where we don't support the creation of resources, it's managed by the IRootResourceProvider for the moment.

File: extensions/odl/capabilities/src/test/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProviderTest.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+		// test getRootResource(id)
+		IRootResource resource = odlIRootResourceProvider.getRootResource(switch1Resource.getId());
+		Assert.assertEquals("OdlRootResourceProvier capability did not retreive the requested resource.", switch1Resource, resource);
+		resource = odlIRootResourceProvider.getRootResource(switch2Resource.getId());
+		Assert.assertEquals("OdlRootResourceProvier capability did not retreive the requested resource.", switch2Resource, resource);
+		resource = odlIRootResourceProvider.getRootResource(peResource.getId());
+		Assert.assertEquals("OdlRootResourceProvier capability did not retreive the requested resource.", peResource, resource);
+
+	}
```
Comment (isartcanyameres) [inline extensions/odl/capabilities/src/test/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProviderTest.java]: I love your tests :)

Comment (adrian-rosello) [inline extensions/odl/capabilities/src/test/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProviderTest.java]: thank you :cry:

File: extensions/odl/capabilities/src/main/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProvider.java (hunk 1 above)
Comment (isartcanyameres) [inline extensions/odl/capabilities/src/main/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProvider.java]: Does it mean that currently our API exposes resources in a different way depending on whether IRootResourceAdministration is available or not?

Comment (isartcanyameres) [inline extensions/odl/capabilities/src/main/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProvider.java]: I see it is.
I've created an issue for tracking this: http://jira.i2cat.net/browse/OPENNAAS-1594

**Pull request merged**